pandas/_libs/lib.pyx

```diff
@@ -581,6 +581,13 @@ def array_equivalent_object(left: object[:], right: object[:]) -> bool:
                     return False
             elif (x is C_NA) ^ (y is C_NA):
                 return False
+            # Only compare scalars to scalars and arrays to arrays
+            elif cnp.PyArray_IsAnyScalar(x) != cnp.PyArray_IsAnyScalar(y):
+                return False
+            # Check if arrays have the same type
+            elif (not (cnp.PyArray_IsPythonScalar(x) or cnp.PyArray_IsPythonScalar(y))
```

Comment:
are you using `not cnp.PyArray_IsPythonScalar(x)` as a standin for `isinstance(x, np.ndarray)`?

Reply:
I don't believe I am. My reasoning for the code is as follows:

The first condition checks if both values are of different types, i.e. scalar vs non-scalar comparison. If they are different then fail.

If both sides are scalars then the current logic in Pandas can correctly handle the comparison. The code still fails though if both sides have a different non-scalar type (e.g. dict vs list comparison). Hence the check to see if we are dealing with non scalars (`not cnp.PyArray_IsPythonScalar(x)`) and then check if they are of the same type.

Reply:
can we not just wrap a try/except around the RichCompareBool? and handle there if there are odd types

this will avoid the perf hit.

Reply:
Put check in a try/except clause. frame_methods.Equals never has a significant perf decrease now (<1.1). Some test from index_cached_properties.IndexCache are sometimes a little bit higher than 1.1 (e.g. 1.11-1.16)

Reply:
ok let's go with this (add a comment as well)

Reply:
Great. I've added the comment to the try/except clause.